### ISSUE TITLE
Alternative to JMeter Samplers

### ISSUE THREAD

**Issue by mchoma** [Hyperfoil/Hyperfoil#226]
This is more question than issue. If this is not apropriate usage, please, tell me where to raise this question better. 

We are evaluating Hyperfoil vs. JMeter. JMeter has big set of Samplers. Sometimes we make use of same of them, e.g. calling shell script [1]. What would be alternative in Hyperfoil? Writing custom step, described in [2] or wrap such action in httpRequest step on server side?

Thank you for you answer in advance.

[1] https://jmeter.apache.org/usermanual/component_reference.html#OS_Process_Sampler
[2] https://hyperfoil.io/quickstart/quickstart8.html

**Comment by rvansa** [comment]
Yes, if you want to execute a shell script/command you will probably want to write custom step. Or maybe rather 2 steps - one that will start the process and store the PID (or process handle) in state, and second that will block (return false) until that process is complete, due to the async nature of scenario execution.

You can find more info on writing extensions in [1] and [2]. While we speak about 'extensions', something as generic as executing a process can be integrated to core. We've recently discussed with @gunnarmorling adding something similar that would trigger remote JMX to start Java Flight Recorder.

Beware though that Hyperfoil is not meant to orchestrate the whole test - build images, start tested system, collect logs etc. There are better tools for that; HF should stay focused on the execution. I understand though that you might want to trigger e.g. an administrative operation somewhere in the middle of the benchmark and don't want to write another server that would control this operation.

[1] https://hyperfoil.io/docs/extensions.html
[2] https://hyperfoil.io/docs/architecture.html

**Comment by mchoma** [comment]
> Beware though that Hyperfoil is not meant to orchestrate the whole test - build images, start tested system, collect logs etc. There are better tools for that; HF should stay focused on the execution.

Totally understand and agree. I really mean shell script execution for load generator not for orchestration. Consider this issue https://issues.redhat.com/browse/JBEAP-22032, we had "double-cklick" reproducer as shell script which worked reliably. I was not able to rewrite it to http requests using jmeter neither.

I have those cases in mind.

But I understand some sort of script orchestration is possible even today using https://hyperfoil.io/userguide/hooks.html, right?
